os/board/rtl8730e/src/component/bluetooth/driver/platform/amebad2/hci/hci_platform.c

```diff
@@ -924,8 +936,54 @@ static uint32_t hci_platform_parse_patch(uint8_t *p_buf, HCI_PATCH_NODE *p_patch
 }
 
 #ifdef CONFIG_SECOND_FLASH_PARTITION
-extern uint8_t *rtlbt_fw_ptr;
-extern uint32_t rtlbt_fw_size;
+static void external_flash_read_stream(struct spi_dev_s *spi, char *rx_buffer, uint32_t length, uint32_t offset)
+{
+    char cmd[EXTERNAL_FLASH_READ_CMD_SIZE] = {0};
+    /* transmit read cmd */
+    cmd[0] = 0x13;
+    cmd[1] = (offset >> 24) & 0xFF;
+    cmd[2] = (offset >> 16) & 0xFF;
+    cmd[3] = (offset >> 8) & 0xFF;
+	cmd[4] = offset & 0xFF;
+
+    SPI_EXCHANGE(spi, (char *)cmd, (char *)rx_buffer, length + EXTERNAL_FLASH_READ_CMD_SIZE);
+    memmove(rx_buffer, rx_buffer + EXTERNAL_FLASH_READ_CMD_SIZE, length);
+}
+
+static uint8_t* hci_platform_get_btfw_patch(uint32_t *rtlbt_fw_size)
+{
+    /* SPI Initialize */		
+    struct spi_dev_s *spi;
+    uint8_t *rtlbt_fw_ptr = NULL;
+    spi = up_spiinitialize(1);
+    SPI_LOCK(spi, true);
+    SPI_SETBITS(spi, 8);
+    SPI_SETFREQUENCY(spi, 1000000);
+    uint8_t *rtlbt_fw_hdr_ptr = NULL;
+    uint32_t rtlbt_fw_addr = 0;
+
+    /* SPI Read size*/
+    rtlbt_fw_hdr_ptr = osif_mem_alloc(RAM_TYPE_DATA_ON, EXTERNAL_FLASH_BINARY_HEADER_SIZE);
+    SPI_SELECT(spi, 1, true);
+    external_flash_read_stream(spi, (char *)rtlbt_fw_hdr_ptr, EXTERNAL_FLASH_BINARY_HEADER_SIZE, HCI_PATCH_ADDR_OFFSET);
+    SPI_SELECT(spi, 1, false);
+    rtlbt_fw_addr = (rtlbt_fw_hdr_ptr[12]) | (rtlbt_fw_hdr_ptr[13] << 8) | (rtlbt_fw_hdr_ptr[14] << 16) | (rtlbt_fw_hdr_ptr[15] << 24);
+    *rtlbt_fw_size = (rtlbt_fw_hdr_ptr[8]) | (rtlbt_fw_hdr_ptr[9] << 8) | (rtlbt_fw_hdr_ptr[10] << 16) | (rtlbt_fw_hdr_ptr[11] << 24);
+    /* Below condition is minimal requirement to meet */
+    DEBUGASSERT((*rtlbt_fw_size > 0 && *rtlbt_fw_size < 0x20000) && (rtlbt_fw_addr == EXT_FLASH_BASE_ADDRESS));
+    if (rtlbt_fw_hdr_ptr) {
+    	osif_mem_free(rtlbt_fw_hdr_ptr);
+    }
+    /* SPI Read Data back */
+    rtlbt_fw_ptr = osif_mem_alloc(RAM_TYPE_DATA_ON, *rtlbt_fw_size);
+    SPI_SELECT(spi, 1, true);
+    /* The binary header size is 32bytes, so we need to add back the offset */
+    external_flash_read_stream(spi, (char *)rtlbt_fw_ptr, *rtlbt_fw_size, rtlbt_fw_addr - (EXT_FLASH_BASE_ADDRESS + HCI_PATCH_ADDR_OFFSET) + 32);
```

Comment:
`external_flash_read_stream` is valid when only CONFIG_SECOND_FLASH_PARTITION is enabled.
If CONFIG_SECOND_FLASH_PARTITION is disabled, is binary made well with internal location and is it working properly?
If not, please fix them because we support both aidual and ai lite

Reply:
If the external flash is not existing, the BT FW patch will be included in the kernel binary.

Reply:
Hi @jeongarmy , regarding your question, it is included in the PR here: https://github.com/Samsung/TizenRT/pull/6190.
I hope it helps.